File: Day23/maximum-score-from-removing-substrings.cc
class Solution {
// Time: O(N)  ::  Space: O(N)
public:
    int maximumGain(string s, int x, int y) {
        auto removeSubString = [](string word, string target) {
            string stack;
            int totalRemoved = 0;
            for (char ch : word) {
                if (!stack.empty() && ch == target[1] && stack.back() == target[0]) {
                    stack.pop_back();
                    totalRemoved += 2;
                } else {
                    stack.push_back(ch);
                }
            }
            return make_pair(stack, totalRemoved);
        };

        int maxScore = 0;
        string firstPair = (x > y) ? "ab" : "ba";
        string secondPair = (x > y) ? "ba" : "ab";

        auto [str1, count1] = removeSubString(s, firstPair);
        maxScore += (count1 / 2) * max(x, y);

        auto [str2, count2] = removeSubString(str1, secondPair);
        maxScore += (count2 / 2) * min(x, y);

        return maxScore;
    }
};




class Solution {
// Time: O(N)  ::  Space: O(1)
public:
    int removeSubString(vector<char>& s, string target) {
        int write = 0, count = 0;
        for (int read = 0; read < s.size(); ++read) {
            s[write++] = s[read];
            if (write > 1 && s[write - 2] == target[0] && s[write - 1] == target[1]) {
                write -= 2;
                count++;
            }
        }
        s.resize(write);
        return count;
    }

    int maximumGain(string s, int x, int y) {
        vector<char> chars(s.begin(), s.end());
        int maxScore = 0;
        if (x > y) {
            maxScore += x * removeSubString(chars, "ab");
            maxScore += y * removeSubString(chars, "ba");
        } else {
            maxScore += y * removeSubString(chars, "ba");
            maxScore += x * removeSubString(chars, "ab");
        }
        return maxScore;
    }
};

File: Day23/maximum-score-from-removing-substrings.js
class Solution {
    // Time: O(N)  ::  Space: O(N)
    maximumGain(s, x, y) {
        const removeSubString = (word, target) => {
            const stack = [];
            let totalRemoved = 0;

            for (const ch of word) {
                if (stack.length > 0 && ch === target[1] && stack[stack.length - 1] === target[0]) {
                    stack.pop();
                    totalRemoved += 2;
                } else {
                    stack.push(ch);
                }
            }
            return [stack.join(""), totalRemoved];
        };

        let maxScore = 0;
        const firstPair = x > y ? "ab" : "ba";
        const secondPair = x > y ? "ba" : "ab";

        const [str1, count1] = removeSubString(s, firstPair);
        maxScore += Math.floor(count1 / 2) * Math.max(x, y);

        const [str2, count2] = removeSubString(str1, secondPair);
        maxScore += Math.floor(count2 / 2) * Math.min(x, y);

        return maxScore;
    }
}




class Solution {
    // Time: O(N)  ::  Space: O(1)
    removeSubString(arr, target) {
        let write = 0, count = 0;
        for (let read = 0; read < arr.length; read++) {
            arr[write++] = arr[read];
            if (write > 1 && arr[write - 2] === target[0] && arr[write - 1] === target[1]) {
                write -= 2;
                count++;
            }
        }
        arr.length = write;
        return count;
    }

    maximumGain(s, x, y) {
        const arr = [...s];
        let maxScore = 0;
        if (x > y) {
            maxScore += x * this.removeSubString(arr, "ab");
            maxScore += y * this.removeSubString(arr, "ba");
        } else {
            maxScore += y * this.removeSubString(arr, "ba");
            maxScore += x * this.removeSubString(arr, "ab");
        }
        return maxScore;
    }
}

